src/server/storage_engine/block_engine/allocator.rs

```diff
@@ -2,54 +2,311 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
-use std::sync::Arc;
-
 use libc::ioctl;
 use nix::fcntl::{open, OFlag};
-use parking_lot::Mutex;
+use std::collections::HashSet;
+
+use crossbeam_queue::ArrayQueue;
 
+use super::AllocatorEntry;
 use crate::server::EngineError;
+extern crate num_cpus;
 
-//#define BLKGETSIZE _IO(0x12,96)	/* return device size /512 (long *arg) */
+//#define BLKGETSIZE _IO(0x12,96)   /* return device size /512 (long *arg) */
 const BLOCKGETSIZE: u64 = 0x1260;
 
 pub const CHUNK: u64 = 512 * 8;
 const SECTOR: u64 = 512;
 
 pub(crate) trait Allocator {
     fn new(path: &str) -> Self;
-    fn allocator_space(&self, lenth: u64) -> u64;
+    fn allocator_space(&self, lenth: u64) -> Vec<AllocatorEntry>;
+    fn release_space(&self, lenth: u64, begin: u64) -> bool;
 }
 
 /*
- * This Allocator use for memory.
+ *This Allocator use for memory.
  */
 #[allow(unused)]
-pub(crate) struct BitmapAllocator {
-    block_space: Arc<Mutex<u64>>,
-    total_aspce: u64,
+pub(crate) struct BlockAllocator {
+    block_group_list: ArrayQueue<BlockGroup>,
+}
+
+struct BlockGroup {
+    block_space_point: u64,
+    block_space: u64,
+    slice_space_manager: Vec<HashSet<u64>>,
```

Comment:
Maybe we need a thread-safe structure instead of Hash-set.

Reply:
Concurrency is guaranteed by arrayqueue